subscribe.go

```diff
@@ -85,3 +93,60 @@ func RegisterSubscriber(scheme string, factory SubscriberFactory) {
 	}
 	subReg[scheme] = factory
 }
+
+// ----------------------------------------------------------------------------
+
+// InMemSubscriber is a subscriber, that consumes messages from seeded data.
+// It is useful mainly for testing.
+type InMemSubscriber struct {
+	mu   sync.Mutex
+	msgs map[string][]SubMessage
+}
+
+// NewInMemSubscriber returns new subscriber, that consumes messages from seeded data.
+func NewInMemSubscriber(messagesByTopic map[string][]SubMessage) *InMemSubscriber {
+	return &InMemSubscriber{
+		msgs: messagesByTopic,
+	}
+}
+
+// Subscribe subscribes to in-memory messages by topic.
+func (s *InMemSubscriber) Subscribe(ctx context.Context, topic string, handler Handler) error {
+	for {
+		// check ctx/cancel BEFORE shift-ing each message:
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		msg, ok := s.shiftMessage(topic)
+		if !ok {
+			return nil
+		}
+
+		if err := handler.Handle(msg); err != nil {
+			return err
+		}
+	}
+}
+
+func (s *InMemSubscriber) shiftMessage(topic string) (SubMessage, bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	msgs := s.msgs[topic]
+
+	if len(msgs) == 0 {
+		return nil, false
+	}
+
+	s.msgs[topic] = msgs[1:]
```

Comment:
I don't like it since it's a recipe for bad memory leaks. you should add a `map[string]int offsets` to store consumed indices

Reply:
Well... My motivation is that it's mostly a test helper, so tried to keep it simple.
It's meant to be drained (usually) and then nil-ed / go out of scope and be collected after use - as it's a seed-only object (messages are not added dynamically, it's not linked with publisher at all).
Should I bother with offsets map?

Reply:
well, positions not offsets really. "shifting" is really a bad practice in Go, I am OK with it in tests but not in general

Reply:
Will migrate to `map[string]struct{ messages []SubMessage; offset int }` or similar approach - better copy seeded data not to affect anything at all.

Reply:
Actually, just should be Ok for such a test-helper.